Fix #631: Implements Thimble Console Back End.

raygervais opened the pull request:
This is the first step toward implementing the suggested [Javascript console](https://github.com/mozilla/thimble.mozilla.org/issues/1675)

Pomax commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -0,0 +1,29 @@
+    var Commands = brackets.getModule("command/Commands");
+
+    var ConsoleManagerRemote = require("text!lib/ConsoleManagerRemote.js");
+    var IFrameBrowser = require("lib/iframe-browser");
```
unused var?

raygervais replied:
Removed in upcoming commit, thank you.

Pomax commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -0,0 +1,29 @@
+    var BrambleEvents = brackets.getModule("bramble/BrambleEvents");
+    var EditorManager = brackets.getModule("editor/EditorManager");
+    var CommandManager = brackets.getModule("command/CommandManager");
+    var Commands = brackets.getModule("command/Commands");
```
unused var?

raygervais replied:
Removed in upcoming commit, thank you.

Pomax commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -0,0 +1,29 @@
+
+    var BrambleEvents = brackets.getModule("bramble/BrambleEvents");
+    var EditorManager = brackets.getModule("editor/EditorManager");
+    var CommandManager = brackets.getModule("command/CommandManager");
```
unused var?

raygervais replied:
Removed in upcoming commit, thank you.

Pomax commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -0,0 +1,29 @@
+    "use strict";
+
+    var BrambleEvents = brackets.getModule("bramble/BrambleEvents");
+    var EditorManager = brackets.getModule("editor/EditorManager");
```
unused var?

raygervais replied:
Removed in upcoming commit, thank you.

Pomax commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -0,0 +1,29 @@
+define(function (require, exports, module) {
+    "use strict";
+
+    var BrambleEvents = brackets.getModule("bramble/BrambleEvents");
```
unused var?

raygervais replied:
Removed in upcoming commit, thank you.

Pomax commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -0,0 +1,29 @@
+    var isConsoleEnabled = false;
+
+    function getRemoteScript() {
+        return "<script>\n" + ConsoleManagerRemote + "</script>\n";
```
This qualifies as extremely unsafe JS, why can this code not be required in the same way as everything else?

raygervais replied:
Apologies. 
I'm working right now with using the same way as many other files for requiring a file, but it returns undefined in the <script> injection of that iframe. Suggestions? 
```js
var ConsoleManagerRemote = require("lib/ConsoleManagerRemote");
```
![image](https://cloud.githubusercontent.com/assets/14265337/23635305/e548f054-029c-11e7-9629-b30c12c6af64.png)

Pomax replied:
Yes, the idea I had was not to load it "as text", but instead make it load the usual way, so that you end up generating a `<script src="lib/ConsoleManagerRemote.js"></script>` in some way. After all, if the code can run "as string", it should be able to run as a normal script `src`, too. The difference being that when you load it as resource from a URL, that drives down the payload size because the browser can cache the file and doesn't need to redownload it every time, cutting down on load time.

humphd replied:
So @Pomax is giving good advice, but in this case we have to do this a hacky way.  We are preparing the injected code that we put into the preview iframe, and we need the raw source of your remote console script so we can concatenate it to the front of the HTML we're going to preview.

You need to require the text value vs. the module as such, and we do it like this:

https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/LinkManager.js#L10

with requirejs' text plugin.

raygervais replied:
Thank you @humphd for the clarification, I had saw 

So, revert back to the implementation in the current PR?

Pomax commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,64 @@
+    //TODO: add support for console.log(one,two, three) arguments
+
+    // Returns a function which calls the specific function in scope
+    var bind = function(func, scope, args) {
```
This is already a function in JavaScript, what does this do differently from the regular `bind()`? And then specifically: can that be documented in the code here, because that's the important information that should be in the function documentation.

That said, is `bind` used anywhere? I don't see it referred to anywhere else in this file.

raygervais replied:
You are absolutely right, it was not being used anywhere. It was written while I was migrating a Console shim suggested, but I forgot entirely to delete it since the shim did not need the aforementioned ```bind()``` function to properly work. Removed in upcoming commit.

Pomax commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,64 @@
+
+    if (!console.assert) {
+        console.assert = function() {
+            var args = Array.prototype.slice.call(arguments, 0);
```
`Array.from(arguments).slice()` should do the trick

raygervais replied:
Changed to your suggestion, much more concise. 
Thank you

raygervais commented:
@Pomax @humphd , I've updated the code to reflect the changes discussed.

humphd commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -0,0 +1,21 @@
+
+    function consoleRequest(msg) {
+        console.log("Bramble Console: " + msg);
+        //alert(msg);
```
Please get rid of this line.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,53 @@
+(function(transport) {
+    "use strict";
+
+    //TODO: add support for console.log(one,two, three) arguments
```
Let's file a new issue on this, and refer back to this PR.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,53 @@
+    //TODO: add support for console.log(one,two, three) arguments
+
+    function _log(s){
+        //See Note below about fixing transport for 'data'
```
Please add a space here:

```js
 // See Note...
```

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,53 @@
+            var args = Array.from(arguments).slicer();
+            var expr = args.shift();
+            if (!expr) {
+                args[0] = " Assertion Failed: " +args[0];
```
Add space after `+`

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,53 @@
+
+    window.console = console;
+
+    // TODO: add support for other methods in console
```
New issue here too please.

raygervais commented:
I created a new issue reflecting the next stages of implementation, and I've fixed your requested changes @humphd.

humphd commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -0,0 +1,21 @@
+        return msg.match(/^bramble-console/);
+    }
+
+    function consoleRequest(msg) {
```
Rename this to `handleConsoleRequest`

raygervais replied:
Changed to `handleConsoleRequest` in upcoming commit.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,49 @@
+
+    // Create Console if not present
+    if (!window.console) {
+        window.console = ({});
```
You don't need the `()` around `{}`.

raygervais replied:
This has been changed as requested in the newest commit.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,49 @@
+(function(transport) {
```
Modify the first last lines to pass the `window.console` object in:

```js
(function(transport, global, console) {
    ...
}(window._Brackets_LiveDev_Transport, window, window.console));
```

raygervais replied:
This has been changed as requested in the newest commit.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    // Create Console if not present
+    if (!window.console) {
```
You don't need this anymore, see http://caniuse.com/#search=console

raygervais replied:
This has been changed as requested in the newest commit. 
Useful website too, thanks for linking that!

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,49 @@
+        window.console = ({});
+    }
+
+    var console = (window.console);
```
We don't need this if we pass `console` in with the outer function.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,49 @@
+    var console = (window.console);
+
+    // Implement Console Log
+    if (!console.log) {
```
You can drop this, it will have `.log` for sure.

raygervais replied:
This has been changed as requested in the newest commit.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,49 @@
+    console.log = _log;
+
+    // Implement other Log Levels to console.log
+    if (!console.debug) {
```
For all these other `console.*` methods we *always* want to re-bind them to `console.log` and therefore `_log` so drop all these `if` checks.

raygervais replied:
This has been changed as requested in the newest commit.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,49 @@
+        };
+    }
+
+    window.console = console;
```
`global.console = console;`

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,29 @@
+    console.info = console.log;
+    console.warn = console.log;
+
+    if (!console.assert) {
```
Drop this, we'll always replace whatever is there (even if nothing).

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,29 @@
+            var expr = args.shift();
+            if (!expr) {
+                args[0] = "Assertion Failed: " + args[0];
+                console.error.apply(console,args);
```
Space after comma.

humphd commented on src/extensions/default/bramble/lib/PostMessageTransport.js:
```diff
@@ -90,7 +91,11 @@ define(function (require, exports, module) {
+                ConsoleManager.handleConsoleRequest(msgObj.data);
+                return;
+            }
+            //trigger message event 
```
Newline above this.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,29 @@
+(function(transport, global, console) {
+    "use strict";
+
+    function _log(s){
```
We need to deal with the fact that the console can take any number of args, and any type of arg--simply saying `s` here is not going to work.  Probably the best thing you can do is to send along an array of all arguments:

```js
function _log() {
  transport.send("bramble-console", Array.from(arguments).slice());
}
```

Now on the receiving side, when you `handleConsoleRequest(msgObj.data)` you'll need to deal with the fact that you have an array of things to pass to the console log:

```js
function handleConsoleRequest(args) {
    // Add an identifier to the front of the args list, so we know this is from us in console
    args = args.unshift("[Bramble Console]");
    // Call console.log, passing our long list of arguments
    console.log.apply(console, args);
}
```

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,29 @@
+            }
+        };
+    }
+
```
I still think you should go look at all the other `console.*` methods and add them, see https://github.com/firebug/firebug-lite/blob/master/content/firebug1.4/consoleInjector.js#L260 and https://developer.mozilla.org/en/docs/Web/API/console

Pomax commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -11,8 +11,39 @@ define(function (require, exports, module) {
+            case "clear":
+                console.clear.apply(console);
+                break;
+            case "time":
```
could we simplify this as:
```
switch(type) {
  case "time":
  case "timeEnd":
    args[0] = args[0] + " " + args[1];
    console[type].apply(console, args);
    break;
  default:
    console[type].apply(console, args);
}
```

raygervais replied:
This has been simplified in the upcoming commit.

Pomax commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -1,28 +1,70 @@
+        }
     }
+    
+    // Implement console.log replacement
```
It feels like here, too, we can simplify by doing some like:
```
var obj = {};

var routeData = generateRouter(fname) {
  return function() {
    var args = Array.from(arguments).slice();
    transportSend(args, fname);
  };
};

["log", "error", "...", ...].forEach( function(fname) {
  obj[fname] = generateRouter(fname);
});
```

raygervais replied:
@Pomax, I've tried to simplify by using yours as an example, and the only way which allows for the console functions to still register is the following:
```js
    function generateRouter(fname) {
        var args = Array.from(arguments).slice();
        transportSend(args, fname);
    }
    
    // Implement console.log replacement
    console.log = function() {
        generateRouter("log");
    }
```
That being said, the `args` sent are `fname` alone:
```js
console.log("Hello Developers"); 
```
Would log `log` into the console. Likewise for the other functions.

Using the `forEach` that you suggested above essentially triggers each of the console functions in the array on loading of the file in the browser. Which is why I tried the method I included above. I'm curious if you have a better implementation, or perhaps we should leave the version which though longer, seems to work with the different arguments. 

Again, just speculating so if I have mentioned anything which is incorrect, I'm open to learning from said mistake.

humphd commented:
@raygervais please let @Pomax and me know when this is ready for review again.

raygervais commented:
@humphd, @Pomax the code is ready for another look over. I fixed the syntax issue which caused Travis to fail, and also simplified where I could at @Pomax's request. Included below is how the console currently functions and how I tested said console functions. 

If we are content with the code changes, I'd be welcome to closing both this and the issue #631 and creating a new issue related to the UI of the console. 

![screenshot_20170316_183411](https://cloud.githubusercontent.com/assets/14265337/24021597/1d0d6b4a-0a78-11e7-871d-61fcc0125b24.png)

humphd commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -0,0 +1,29 @@
+
+    function handleConsoleRequest(args, type) {
+        // Add an indentifier to the front of the args list
+        args.unshift("[Bramble Console]:");
```
We can remove this now that we're done testing.

humphd commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -0,0 +1,29 @@
+        args.unshift("[Bramble Console]:");
+       
+        // Time related arguments
+        if(type === "time" || type === "timeEnd") {
```
I think you should handle this case in the remote code, where you're dealing with those methods vs. here, which is really far away from the code that originally caught it.  This code should be fairly agnostic about what to do with the message it's receiving, it should just handle it without modification.

humphd commented on src/extensions/default/bramble/lib/ConsoleManager.js:
```diff
@@ -0,0 +1,29 @@
+            args[0] = args[0] + " " + args[1];
+        }
+        
+        console[type].apply(console, args);
```
Please add a `// TODO: show this in custom console UI, see issue #...`

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,76 @@
+    
+    function transportSend(args, type) {        
+        // To avoid Runtime.evaluate messages
+        if(args[0] !== "Runtime.evaluate") {
```
This is being fixed in https://github.com/mozilla/brackets/pull/628, so you can remove this check altogether.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,76 @@
+(function(transport, global, console) {
+    "use strict";
+    
+    function transportSend(args, type) {        
```
I'm not going to block you on changing this, but I find it surprising that you don't have the `type` first, and `args` second.  That's the opposite of what I'd expect.  Just a comment, in case your choice was random.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,76 @@
+    }
+    
+    function generateRouter(fname) {
+        var args = Array.from(arguments).slice();
```
Let's DRY this code out (Don't Repeat Yourself):

```js
// Replace console methods with our own versions for integrated dev tools logging
["log", "warn", ...].forEach(function(type) {
  console[type] = function() {
    var args = Array.from(arguments).slice();
    transportSend(args, type);
  };
});
```

When you're basically copy/pasting the same code a dozen times, it's way safer to write code that generates the code, since it becomes boilerplate, and it's easy to introduce bugs with a mistake in one of the items.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,76 @@
+        transportSend(args, "clear");
+    };
+    
+    // Implement console.time replacement
```
For `time, timeEnd, assert` which are not using the same boilerplate, you can do this no problem.  However, please deal with your custom arg processing for `time, endTime` in here.

humphd commented on src/extensions/default/bramble/lib/PostMessageTransportRemote.js:
```diff
@@ -38,11 +38,15 @@
+         * @param {string} level The console type used
          */
-        send: function (msgStr) {
+        send: function (msgStr, data, level) {
```
This is an example of a "leaky abstraction."  You have code elsewhere that knows what a `level` is, and here you're exposing that lower level code's concepts.  As you move from the `console` (where `level` has meaning) to the more generic `transport`, where you are only sending data across a channel, you want to hide this detail.

In this case, you should merge the `level` into your `data`, such that you would call things like so:

```js
// before, leaking too much internal abstraction info
transport.send("bramble-console", args, type);

// after, with all data bundled into one object
transport.send("bramble-console", {args: args, type: type});
```

Now you can just send `data` across the wire, and let the `ConsoleManager` deal with the more complex object that has both `args` and `level` attached.  The rest of your code using the transport can be blissfully unaware.

Another side comment: in JS it would be `dataType` but never the pythonic `data_type`.

raygervais commented:
@humphd, Thank you for the code review. I will get to these changes immediately. When I am done, would you like me to branch off while this one awaits code review and work on the Console UI? Or, commit to here?

PS, thank you for clarifying some of the syntactical issues I had previously, I do appreciate learning the most optimal way for Mozilla-esque development and concerns.

humphd commented:
The thing with `git` is that you can fork off of this and start your UI, then just `rebase` later onto master when your backend stuff lands.  You don't need to block.

Don't push any more new code to this branch, do all new stuff (except for fixes) on new PRs and branches.

raygervais commented:
@humphd, I've updated the files with your requested changes. When you get a chance the code is ready for another review, and in the mean time I will keep working using another forked-branch on the user interface.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,34 @@
+        transport.send("bramble-console", data);
+    }
+    
+	// Implement standard console.* functions
```
Some odd indenting here.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,34 @@
+    
+	// Implement standard console.* functions
+	["log"
+        , "warn"
```
This is a style thing, but I'm not a fan of using a leading comma.  Can you please put it at the end of each item?

```js
["log",
  "warn",
```

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,34 @@
+        };
+    });
+    
+    // Replace default console.assert with custom
```
Dump this comment, it doesn't add any info.

raygervais commented:
@humphd, I've fixed all your requested changes, and am ready for another review when you get the chance.

humphd commented on src/extensions/default/bramble/lib/ConsoleManagerRemote.js:
```diff
@@ -0,0 +1,33 @@
+    
+    // Implement standard console.* functions
+    ["log",
+        "warn",
```
Not quite:

```js
    ["log",
     "warn",
     ...,
     "timeEnd"].forEach(function(type) {
```

raygervais replied:
@humphd, fixed!

humphd commented:
Thank you!  Excited to see the UI happen next.

raygervais commented:
Thank you @humphd, @Pomax for the code reviews and stylistic pointers!
